cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -0,0 +1,174 @@
+                                           scan_tile_status status)
+  {
+    auto thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (thread_idx < count) {  //
+      // this is UB if tile_status gets assigned from multiple threads.
+      tile_status[(base_tile_idx + thread_idx) % num_tiles] = status;
+    }
+  }
+
+  __device__ inline void set_partial_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_partial + offset, value);
+    __threadfence();
+    cub::ThreadStore<cub::STORE_CG>(tile_status + offset, scan_tile_status::partial);
+  }
+
+  __device__ inline void set_inclusive_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_inclusive + offset, value);
+    __threadfence();
+    cub::ThreadStore<cub::STORE_CG>(tile_status + offset, scan_tile_status::inclusive);
+  }
+
+  __device__ inline T get_prefix(cudf::size_type tile_idx, scan_tile_status& status)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+
+    while ((status = cub::ThreadLoad<cub::LOAD_CG>(tile_status + offset)) ==
+           scan_tile_status::invalid) {
+      __threadfence();
+    }
+
+    if (status == scan_tile_status::partial) {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_partial + offset);
+    } else {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+    }
+  }
+
+  __device__ inline T get_inclusive_prefix(cudf::size_type tile_idx)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    while (cub::ThreadLoad<cub::LOAD_CG>(tile_status + offset) != scan_tile_status::inclusive) {
+      __threadfence();
+    }
+    return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+  }
+};
+
+template <typename T>
+struct scan_tile_state {
+  rmm::device_uvector<scan_tile_status> tile_status;
+  rmm::device_uvector<T> tile_state_partial;
+  rmm::device_uvector<T> tile_state_inclusive;
+
+  scan_tile_state(cudf::size_type num_tiles,
+                  rmm::cuda_stream_view stream,
+                  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+    : tile_status(rmm::device_uvector<scan_tile_status>(num_tiles, stream, mr)),
+      tile_state_partial(rmm::device_uvector<T>(num_tiles, stream, mr)),
+      tile_state_inclusive(rmm::device_uvector<T>(num_tiles, stream, mr))
+  {
+  }
+
+  operator scan_tile_state_view<T>()
+  {
+    return scan_tile_state_view<T>{tile_status.size(),
+                                   tile_status.data(),
+                                   tile_state_partial.data(),
+                                   tile_state_inclusive.data()};
+  }
+
+  inline void set_seed_async(T const seed, rmm::cuda_stream_view stream)
+  {
+    auto x = tile_status.size();
+    auto y = scan_tile_status::inclusive;
```

Comment:
Is there a deeper meaning why this is named `y`. Having `x` and `y` had put me in the mental state that there's something 2D is going on here initially.
```suggestion
    auto target_status = scan_tile_status::inclusive;
```

Reply:
ah, no, sorry. `set_element_async` doesn't like rvalue refs, so I had to make an lvalue and didn't make a good name for it.